rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNode.java

```diff
@@ -0,0 +1,172 @@
+import com.google.gson.JsonObject;
+import com.nimbusds.oauth2.sdk.util.CollectionUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.thingsboard.common.util.JacksonUtil;
+import org.thingsboard.rule.engine.api.RuleNode;
+import org.thingsboard.rule.engine.api.TbContext;
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.server.common.data.AttributeScope;
+import org.thingsboard.server.common.data.DataConstants;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.id.AssetId;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.kv.AttributeKvEntry;
+import org.thingsboard.server.common.data.msg.TbMsgType;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.data.relation.*;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+@Slf4j
+@RuleNode(type = ComponentType.ACTION,
+        name = "gps multi geofencing events",
+        configClazz = TbGpsMultiGeofencingActionNodeConfiguration.class,
+        relationTypes = {"Success", "Entered", "Left", "Inside", "Outside"},
+        nodeDescription = "Produces incoming messages using GPS based geofencing",
+        nodeDetails = "Extracts latitude and longitude parameters from incoming message and returns different events " +
+                "based on configuration parameters")
+public class TbGpsMultiGeofencingActionNode extends AbstractGeofencingNode<TbGpsMultiGeofencingActionNodeConfiguration> {
+
+    private GeofencingProcessor geofencingProcessor;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        super.init(ctx, configuration);
+        geofencingProcessor = new GeofencingProcessor(ctx, this.config);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        List<EntityId> matchedGeofenceIds = findMatchedZones(ctx, msg);
+        GeofenceResponse geofenceResponse = geofencingProcessor.process(msg.getMetaDataTs(), ctx.getSelfId(), msg.getOriginator(), matchedGeofenceIds);
+        processGeofenceResponse(ctx, msg, msg.getOriginator(), geofenceResponse);
+        ctx.tellSuccess(msg);
+    }
+
+    private void processGeofenceResponse(TbContext ctx, TbMsg originalMsg, EntityId originatorId, GeofenceResponse geofenceResponse) {
+        TbMsgMetaData metaData = originalMsg.getMetaData().copy();
+        metaData.putValue("originatorId", originatorId.toString());
+        metaData.putValue("originatorName", getOriginatorName(ctx, originatorId));
```

Comment:
Why do we need these 2 fields in the message metadata?

Reply:
originator id field in the metadata is used to show which originator has entered a geofence. Message from the node is sent with originator id as geofence, not the original entity.

Reply:
this was requirement from Andrey Tkachenko

Reply:
https://thingsboard-portal.atlassian.net/browse/ATT-2

Reply:
I have removed originator name